lib/Conversion/ExportVerilog/PrepareForEmission.cpp

```diff
@@ -764,6 +764,40 @@ static LogicalResult legalizeHWModule(Block &block,
       continue;
     }
 
+    // Force array index expressions to be a simple name when the
+    // `mitigateVivadoArrayIndexConstPropBug` option is set.
+    if (options.mitigateVivadoArrayIndexConstPropBug &&
+        isa<ArraySliceOp, ArrayGetOp, ArrayIndexInOutOp,
+            IndexedPartSelectInOutOp, IndexedPartSelectOp>(&op)) {
+
+      // Check if the index expression is already a wire.
+      WireOp wireOp;
+      ReadInOutOp readOp;
+      if (auto maybeReadOp =
+              op.getOperand(1).getDefiningOp<sv::ReadInOutOp>()) {
+        if (auto maybeWireOp = maybeReadOp.getInput().getDefiningOp<WireOp>()) {
+          wireOp = maybeWireOp;
+          readOp = maybeReadOp;
+        }
+      }
+
+      // Insert a wire and read if necessary.
+      ImplicitLocOpBuilder builder(op.getLoc(), &op);
+      if (!wireOp) {
+        wireOp =
+            builder.create<WireOp>(op.getOperand(1).getType(), "_ARRAY_IDX");
```

Comment:
I think the convention is to prefix with _GEN, so maybe use _GEN_ARRAY_IDX so we can still know it starts with _GEN?